Add DiscreteNonParametric distribution

GordStephen opened the pull request:
This implements a new generic, discrete, sparse distribution type for representing and sampling from an arbitrary PDF, as was discussed [on Discourse](https://discourse.julialang.org/t/construct-a-generic-discrete-distribution-in-distributions-jl/2238?u=gordstephen) a while ago. As of right now the type is called `Generic`, although there's probably a better name for for it...

If there's interest in merging this, I can add documentation, more robust testing, and any outstanding functionality (`mfg`, `cf`, etc).

andreasnoack commented:
Thanks for contributing this. This one is extremely similar to `Categorical` so it would be great if we could think through if they could either be merged or at least how much code they could share.

GordStephen commented:
Sounds good - they already share the sampler and it should be easy to consolidate most of the other functionality as well. They probably can't be merged since the point of `Categorical` is that you don't need to store a value vector, while the point of `Generic` is that you _want_ to store a value vector - but maybe they could share an abstract parent type (`ArbitraryDistribution <: DiscreteUnivariateDistribution` or similar) with most of the methods defined on that?

andreasnoack commented:
An option might be just to use a unit range for the value vector. It should be very light weight. In my book, this `Generic` distribution is the same as a categorical distribution. However, it might be that we want to separate them for efficiency reasons but I'd prefer if it wasn't necessary.

GordStephen commented:
Agreed, I think that's the same approach as my edit. I can't think of any reasons combining them would need to be any less efficient, although I'll run some performance tests to verify that.

Semantically, to me at least, a categorical distribution suggests that the support is non-numeric, i.e the integers in the support are just proxies for some other label, and their ordering/relative values are somewhat arbitrary. Given that connotation I would be hesitant to label a distribution over a pontentially-quantitatively-meaningful real-valued support as `Categorical`... I don't particularly like `Generic` either, but whatever it's called, I could see switching `Categorical` to be an alias for an instance of the more general case:

```julia
struct Generic{T, P, S <: AbstractVector{T}}  ... end
const Categorical{P} = Generic{Int, P, Base.OneTo{Int}} where P
```

`Generic` could also provide single-argument constructors to maintain the existing `Categorical` API.

jmxpearson commented:
Just a bikeshedding note: what about `Atomic`? In statistics, I've often seen this kind of distribution with support only at fixed values described as a collection of atoms.

Not that "atomic" doesn't have other meanings in software...

GordStephen commented:
Still more work to be done, but I've converted `Categorical{T}` into a type alias for `Generic{Int,T,UnitRange{Int}}`. It's a drop-in replacement - the only test that needed changing to pass was one that accessed the `Categorical.K` field directly rather than using the `ncategories` getter method. I used `UnitRange{Int}` to maximize compatibility with existing code and tests, although eventually I think `Base.OneTo{Int}` would better encode the fact that the support is always 1:K.

As far as bikeshedding the name, a few other options:
 - `NonParametric`, although it could cause confusion with `Empirical`
 - `Arbitrary`, although it's no more descriptive than `Generic`

GordStephen commented:
I think all of the missing methods should be implemented now, but correct me if I've missed anything. As noted in https://github.com/JuliaStats/Distributions.jl/pull/661#issuecomment-329593704, the support of `DiscreteUnivariateDistribution`s are assumed to contain `Int`s, so `Generic` probably shouldn't have that as its supertype....

leclere commented:
Just wanted to state that it would be usefull for me :-)

matbesancon commented:
@GordStephen could you rebase the PR on master before we can move on with this work? There are conflicts

GordStephen commented:
Sure, will do

matbesancon commented:
Thanks!

GordStephen commented:
Ok, this is rebased and updated for 0.7.

GordStephen commented on src/samplers/aliastable.jl:
```diff
@@ -0,0 +1,25 @@
+end
+rand(s::AliasTable) = rand(Random.GLOBAL_RNG, s)
+
+show(io::IO, s::AliasTable) = @printf(io, "AliasTable with %d entries", ncategories(s))
```
This is just a cut/paste from the existing Categorical code - https://github.com/JuliaStats/Distributions.jl/pull/634/files#diff-1eaa9a93e0334a4e936ae1f750fccff1L52 - the contents of the `AliasTable` struct aren't particularly intuitive to look at so I would imagine that's why the `show` method was defined in the first place. But if the preference is to remove the method, that's fine with me.

ararslan replied:
Ah I missed that it's the same as what's already there. Never mind then, don't worry about it.

GordStephen commented:
Thanks for taking a look @ararslan  - in addition to addressing your comments I realized the `suffstats` method with sample weights was untested and had some issues, those are addressed now.

matbesancon commented:
@GordStephen what's the status for this on your side? You did all the modifications you wanted to?

GordStephen commented:
Yep, although I still don't really like the name... As discussed elsewhere (#661 / #681 / #699) the type hierarchy still doesn't properly describe this distribution, so we should think more about that to.

Once this is merged there's probably opportunity to better integrate with EmpiricalUnivariateDistribution / #661 - is an Empirical not just a Generic that retains the original data?

leclere commented:
If I may just point that "Generic" might imply that it also contain continuous distribution (e.g. with analytical pdf) ? 

As was suggested earlier : 
- Atomic
- DiscreteGeneric
could be considered.

matbesancon commented:
do you think there are things we could do in this PR directly, at least for the name

GordStephen commented:
Yes, no point in merging this and then changing the name later. I'm fine with `Atomic` as a name but worry that there's potential for confusion with [`Threads.Atomic`](https://docs.julialang.org/en/stable/base/multi-threading/#Base.Threads.Atomic)... What about `NonParametricDiscrete` / `DiscreteNonParametric`?

jmxpearson commented:
`AtomicDistribution`?

On Mon, Sep 24, 2018 at 3:14 PM Gord Stephen <notifications@github.com>
wrote:

> Yes, no point in merging this and then changing the name later. I'm fine
> with Atomic as a name but worry that there's potential for confusion with
> Threads.Atomic
> <https://docs.julialang.org/en/stable/base/multi-threading/#Base.Threads.Atomic>...
> What about NonParametricDiscrete / DiscreteNonParametric?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaStats/Distributions.jl/pull/634#issuecomment-424091699>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADBM51ywegBI3XU-ILpiRiKfAldmRy6Xks5ueS8VgaJpZM4OSLui>
> .
>

matbesancon commented:
do you think there are things we could do in this PR directly, at least for the name?

jmxpearson commented:
Sorry, I don't follow. I was just voting for something along the lines of `Atomic`. 

Is there something else you're asking?

GordStephen commented:
@jmxpearson I'm assuming @matbesancon's last comment was meant as an edit to his comment before that?

Suffixing with -Distribution is a bit awkward and doesn't follow the naming convention of the other distributions in the package. Are there objections to simply referring to this distribution as non-parametric? `DiscreteNonParametric` would follow the convention set by `DiscreteUniform`?

matbesancon commented:
Sorry for being unclear, I was refering to the term `Generic` which could be a bit misleading:
> Yep, although I still don't really like the name... As discussed elsewhere (#661 / #681 / #699) the type hierarchy still doesn't properly describe this distribution, so we should think more about that too.

Without modifying the whole type hierarchy in this PR, could/should these changes be adapted now

maximerischard commented:
Thank you for implementing this. I also think “Atomic” is a good name (in fact that's what I was googling before finding this PR). Currently, there is a problem with rand() returning non-integers:

```julia
g=Generic([1.0, 3.2, 4.0], [0.2, 0.3, 0.5])
rand(g, 10)
```
```
InexactError: Int64(Int64, 3.2)

Stacktrace:
 [1] Type at ./float.jl:700 [inlined]
 [2] convert at ./number.jl:7 [inlined]
 [3] setindex!(::Array{Int64,1}, ::Float64, ::Int64) at ./array.jl:769
 [4] _rand!(::Distributions.GenericSampler{Float64,Array{Float64,1}}, ::Array{Int64,1}) at $JULIA_DEPOT_PATH/dev/Distributions/src/genericrand.jl:37
 [5] rand(::Generic{Float64,Float64,Array{Float64,1}}, ::Int64) at $JULIA_DEPOT_PATH/dev/Distributions/src/univariates.jl:176
 [6] top-level scope at In[35]:2
```
It seems to me that `Generic` (or `Atomic`, as it may), should not be of abstract type `Discrete`, since discrete normally means the support is integers.

GordStephen commented:
How about `DiscreteAtomic` for a name - or is that obnoxiously redundant?

matbesancon commented:
I like `DiscreteAtomic`. @GordStephen could you add tests for this.
Also we need to document the whole PR a bit more. Could you add docstrings to all new types and functions, and in the `file of the @docs` folder which best suits?

RuiRojo commented:
Hi. Thanks for your work of course.

I don't like anything with `Atomic` here, since I think it's a fancy "discrete", which doesn't say enough. Nor Generic. But I'm not coming up with anything I like either.

Perhaps if it was just an extension of Categorical, with some extra type parameter. Or maybe `GeneralizedCategorical`?

matbesancon commented:
"fancy" is not always bad, at least it's specific and associated with discrete distributions. We have to keep in mind that the literature does not have clear universal terms for these concepts (discrete over a setthat's not in N), so it's fine to come up with new, specific terms for this

RuiRojo commented:
In many words, this is a "general discrete distrubition with finite support", right?

I like the idea of coming up with a new specific term. But it shouldn't already mean something that's in conflict with other existing distributions, such as just `Discrete`/`Atomic`. IIUC, an atomic distribution is the same (if not more general) than a discrete one, so it would be bad if it "derived" from Discrete.

matbesancon commented:
@GordStephen while some thoughts are still given on the naming, could you add some documentation in `docs` and docstring to all new non-trivial functions (it doesn't hurt, even the unexported) :)

GordStephen commented:
Will do, although I'm traveling a fair bit this week and next so there might be some delay.

matbesancon commented:
bumping this one if you have time to add the doc pages & docstring

GordStephen commented:
I've added some docstrings and updated the Univariate docs page, but let me know if there should be updates anywhere else.

I also renamed the distribution to `DiscreteNonParametric` for now, which is at least unambiguous: one thing we can all agree on is that `Generic` isn't a great option.

matbesancon commented on src/samplers/categorical.jl:
```diff
@@ -4,7 +4,7 @@ struct CategoricalDirectSampler <: Sampleable{Univariate,Discrete}
     prob::Vector{Float64}
 
     function CategoricalDirectSampler(p::Vector{Float64})
```
this could be made more generic here:
```
function CategoricalDirectSampler(p::V) where {V<:AbstractVector{<:Real}}
```

GordStephen replied:
Good point, I think more generally it would be natural to address #743 in this PR

matbesancon commented:
Agreed, if you can make the required modifications

On Mon, Oct 29, 2018, 23:13 Gord Stephen <notifications@github.com> wrote:

> *@GordStephen* commented on this pull request.
> ------------------------------
>
> In src/samplers/categorical.jl
> <https://github.com/JuliaStats/Distributions.jl/pull/634#discussion_r229118001>
> :
>
> > @@ -4,7 +4,7 @@ struct CategoricalDirectSampler <: Sampleable{Univariate,Discrete}
>      prob::Vector{Float64}
>
>      function CategoricalDirectSampler(p::Vector{Float64})
>
> Good point, I think more generally it would be natural to address #743
> <https://github.com/JuliaStats/Distributions.jl/issues/743> in this PR
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaStats/Distributions.jl/pull/634#discussion_r229118001>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHRRslaUha22BT8Yxihqf0qt8zVucmRpks5up32ngaJpZM4OSLui>
> .
>

GordStephen commented:
Ok, non-`Vector` probability containers should now be supported for `DiscreteNonParametric` and therefore also `Categorical`. I also switched the support of `Categorical` over to be `Base.OneTo{Int}` instead of just `UnitRange{Int}` - happily the only test breakage occured in a test that was explicitly looking for a `UnitRange` type (switching to `AbstractUnitRange` solved it).

matbesancon commented:
thanks a lot @GordStephen 
merging in 24 hours if no additional comments
cc @ararslan @simonbyrne

GordStephen commented:
Doing some final checks, I realized the tests for this distribution didn't include `test_distr`. Unfortunately there are several assumptions about uniformly-spaced integer support for discrete distributions in that function that prevent it from working properly on this distribution, but I did make what noninvasive changes I could to allow most of the tests there to run (which yielded a few minor required changes, now addressed).

matbesancon commented:
Thanks, merged

andreasnoack commented:
A bit late here but what I suggested in https://github.com/JuliaStats/Distributions.jl/pull/634#issuecomment-314556585 was that we just used the name `Categorical` for this distribution unless it had to be a separate distribution for efficiency reasons. In the end, we ended up using the same distribution for both but invented a new name instead of just using `Categorical`.

I think we should revisit the name since I believe `DiscreteNonParametric` is a misnomer. What is non-parametric about this distribution? I'll suggest that we rename it to `Categorical`. We can keep the existing `Categorical` that doesn't take a sample space argument.

nalimilan commented:
The fact that `Categorical` is restricted to integers in `1:k` is indeed confusing. It would be simpler to allow specifying particular values as an option.

GordStephen commented:
To me the main distinction between `DiscreteNonParametric` (happy to revisit the name) and `Categorical` should be that the support of `DiscreteNonParametric` is numeric and so has meaningfully defined moments and other quantitative properties, whereas a `Categorical` distribution can have an arbitrary and unordered support (e.g. `["apples", "bananas", "coconuts"]`). So while I'm certainly in favor of allowing `Categorical` to have non-integer support, I wouldn't want to just rename `DiscreteNonParametric` to `Categorical`.

To the point about whether this distribution is non-parametric, I suppose one could argue that it has parameters, just lots of them (a probability and value for each element of the support). That said, a distribution defined by a kernel density estimate has similar properties, and KDEs are considered to be defined non-parametrically.
